plugin/db/driver.go

```diff
@@ -332,6 +332,14 @@ type Driver interface {
 	ExecuteMigration(ctx context.Context, m *MigrationInfo, statement string) (int64, string, error)
 	// Find the migration history list and return most recent item first.
 	FindMigrationHistoryList(ctx context.Context, find *MigrationHistoryFind) ([]*MigrationHistory, error)
+	// InsertPendingMigration inserts a PENDING migration record.
+	InsertPendingMigration(ctx context.Context, tx *sql.Tx, m *MigrationInfo, sequence int, statement string, prevSchema string) (int64, error)
+	// MarkMigrationAsDone marks the last inserted migration record to be DONE.
+	MarkMigrationAsDone(ctx context.Context, tx *sql.Tx, duration int64, insertedID int64, updatedSchema string) error
+	// MarkMigrationAsFailed marks the last inserted migration record to be FAILED.
+	MarkMigrationAsFailed(ctx context.Context, tx *sql.Tx, duration int64, insertedID int64) error
+	// TablePrefix returns the prefix to be added to the migration history table.
+	TablePrefix() string
```

Comment:
I like the idea of splitting them into smaller methods. How about making them as private method since they are not intended to be called by other modules (e.g. calling MarkMigrationAsDone by external module is an anomaly and probably won't work) ?

Reply:
I'm afraid we can't do that in a Go interface.

Reply:
NM:)

Reply:
Then I see there are tradeoffs between 2 approaches since leaking the private method also has bad smells. Let @spinningbot weigh in since he refactored the last time

Reply:
BTW, if you only have time to setup and write tests for one of the drivers, which one would you choose? MySQL?

Reply:
Yes, MySQL, then PostgresSQL

Reply:
I have come up with a design that could encapsulate the different behaviors, and at the same time doesn't leak too much implementation detail. But the price is to introduce yet another concept into the `db` package, a `migrationRecorder`.

1. We make this a private interface so that code outside this package has no access to it;
2. For each driver implementation there's a corresponding `migrationRecorder` implementation, that's where we put the different implementations;
3. Lookup the `migrationRecorder` using `dbType`.

What do you think?

Reply:
migrationRecorder implements the 3 method we extract here?

Reply:
Yes, does it sound unnatural in this domain?

Reply:
Yeah, I prefer the new approach since it's an implementation detail and won't complicate the external interface.

Reply:
Seems like it's impossible either, because `driver.go`, `driverutil.go` and the driver implementations all live in different packages.

Reply:
Yup, then I will probably not go for this route. I still prefer to keep the external interface sane. On the other hand, there could be still some room to improve the readability in how we arrange those 3 steps.